### PR TITLE
[inductor] Insert triton barrier before storing to inplace buffers

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100769

The linked issue demonstrates a triton bug where a load broadcasted
over multiple warps may see the result of a store that happens later
in the triton program. The workaround is to add a barrier before
storing, which enforces that all warps have already read the data.

e.g. in `test_embedding_var_mean` we now generate:
```python
    tl.debug_barrier()
    tl.store(in_out_ptr1 + (tl.broadcast_to(x0, [XBLOCK, 1])), tmp17, None)
```

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by jansel — APPROVED** [review]
Any perf impact here?

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1167,6 +1167,19 @@ def store(self, name, index, value, mode=None):
...
+        # Guard against write-after-read corruption in triton.
+        # See # https://github.com/openai/triton/issues/1615
+        # This triton bug means that a load which is broadcasted over multiple
+        # warps may see the result of a store that happens later in the triton
+        # program. The workaround is to add a barrier before storing, which
+        # enforces that all warps have already read the data.
+        is_inplace = name in self.args.inplace_buffers
+        is_broadcasted = not set.issubset(
+            set(self.range_tree_nodes.keys()), original_index.free_symbols
+        )
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1179]: It'd be nice to have this in `utils.py`. I need exactly this for another PR.

Comment (peterbell10) [reply torch/_inductor/codegen/triton.py:1184]: @lezcano I hope you don't mind I'll just merge this as is and I'll trust you to refactor it.

Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1184]: for sure

**Pull request closed without merging**